viewing_party/party.py

```diff
@@ -1,23 +1,193 @@
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+# method 1 
+def get_available_recs(user_data):
+    # Step 1: Get friends' unique watched movies using Wave 3
+    friends_unique_watched = get_friends_unique_watched(user_data)
+
+    # Step 2: From these movies, filter out those hosted on platforms the user subscribes to
+    subscriptions = user_data["subscriptions"]
+    recommended_movies = []
+
+    for movie in friends_unique_watched:
+        if movie["host"] in subscriptions:
+            recommended_movies.append(movie)
+
+    return recommended_movies
+
+# method 2
+def get_available_recs(user_data):
+    # Step 1: Get a list of movies that friends have watched, with no duplicate titles
+    friend_watched_list = set()
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friend_watched_list.add(movie["title"])
+    
+    # Step 2: Filter out movies that the user has already watched
+    user_watched_list = set()
+    for movie in user_data["watched"]:
+        user_watched_list.add(movie["title"])
+
+    friend_watched_user_not_list = []
+    for title in friend_watched_list:
+        if title not in user_watched_list:
+            friend_watched_user_not_list.append(title)
+
+    # Step 3: From the remaining movies, find those hosted on platforms the user subscribes to
+    movie_in_host = []
+    for friend in user_data ["friends"]:
+        for movie in friend["watched"]:
+            if (movie["title"] in friend_watched_user_not_list 
+                and movie["host"] in user_data["subscriptions"] 
+                and movie not in movie_in_host):
+                movie_in_host.append(movie)
+    return movie_in_host
 
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+
+    most_frequent_genre = get_most_watched_genre(user_data)
+
+    #Determine a list of recommended movies. 
+    recommended_movies = []
+
+    #List of titles the user has already watched
+    user_watched_titles = [movie["title"] for movie in user_data["watched"]]
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if (movie["title"] not in user_watched_titles and
+                movie["genre"] == most_frequent_genre and
+                movie not in recommended_movies):
+                recommended_movies.append(movie)
+    return recommended_movies
+
+def get_rec_from_favorites(user_data):
+    favorites = user_data["favorites"]
+    friends_watched_titles = []
+    recommended_movies = []
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_watched_titles.append(movie["title"])
+    
 
+    for movie in favorites:
+        if (movie["title"]) not in friends_watched_titles:
+            recommended_movies.append(movie)
+    
+    return recommended_movies
```

Comment:
A lot of the logic in this function is dealing with excluding movies the friends have watched. What if we started with a list of movies only the user has watched (we have a helper from a previous wave for this!), and filtered that list to only those in the favorites list?

That would make this function more concise!